src/main/java/edu/rit/se/testsmells/testsmell/TestSmellDetector.java

```diff
@@ -0,0 +1,100 @@
+import org.apache.commons.lang3.StringUtils;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class TestSmellDetector {
+
+    private List<InputStream> inputStreams;
+    private List<AbstractSmell> testSmells;
+
+    /**
+     * Instantiates the various test smell analyzer classes and loads the objects into an List
+     */
+    public TestSmellDetector() {
+        testSmells = new ArrayList<>();
+        inputStreams = new ArrayList<>();
+    }
+
+    public void addDetectableSmell(AbstractSmell smell) {
+        testSmells.add(smell);
+    }
+
+    public void clear() {
+        for (InputStream inputStream : inputStreams) {
+            try {
+                inputStream.close();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+        for (AbstractSmell smell : testSmells) {
+            smell.clear();
+        }
+        testSmells.clear();
+    }
+
+    /**
+     * Factory method that provides a new instance of the TestSmellDetector
+     *
+     * @return new TestSmellDetector instance
+     */
+    public static TestSmellDetector createTestSmellDetector() {
+        return new TestSmellDetector();
+    }
+
+    /**
+     * Provides the names of the smells that are being checked for in the code
+     *
+     * @return list of smell names
+     */
+    public List<String> getTestSmellNames() {
+        return testSmells.stream().map(AbstractSmell::getSmellName).collect(Collectors.toList());
+    }
+
+    /**
+     * Loads the java source code file into an AST and then analyzes it for the existence of the different types of test smells
+     */
+    public void detectSmells(TestFile testFile) {
+        CompilationUnit testFileCompilationUnit = parseIntoCompilationUnit(testFile.getTestFilePath());
+
+        CompilationUnit productionFileCompilationUnit = parseIntoCompilationUnit(testFile.getProductionFilePath());
+
+        for (AbstractSmell smellPrototype : testSmells) {
+            AbstractSmell smell = smellPrototype.recreate();
+            try {
+                smell.runAnalysis(testFileCompilationUnit, productionFileCompilationUnit, testFile.getTestFileNameWithoutExtension(), testFile.getProductionFileNameWithoutExtension());
+            } catch (FileNotFoundException ignored) {
+            }
+
+            testFile.addDetectedSmell(smell);
+        }
+
+    }
+
+    private CompilationUnit parseIntoCompilationUnit(String filePath) {
+        if (StringUtils.isEmpty(filePath)) {
+            return null;
+        }
+        InputStream testFileInputStream = null;
+        try {
+            testFileInputStream = new FileInputStream(filePath);
+        } catch (IOException e) {
+            testFileInputStream = getClass().getResourceAsStream(filePath);
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [UI: Usage of GetResource may be unsafe if class is extended (UI_INHERITANCE_UNSAFE_GETRESOURCE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#ui-usage-of-getresource-may-be-unsafe-if-class-is-extended-ui-inheritance-unsafe-getresource).
Calling this.getClass().getResource(...) could give results other than expected if this class is extended by a class in another package.

Reply:
Should making TestSmellDetector a final class be enough? This problem only occurs when both TestSmellDetector is extended and a relative path is used (not relative to the current directory, but relative to the class' resources path). 

I don't think it's a critical problem, but I'm open to other suggestions anyways.

Reply:
Thank you for your quick reply.  Since this is a test class, this problem is not critical. :)